### PR TITLE
Reuse binaries from compiler directory to reduce SDK size

### PR DESCRIPTION
**Pull request by captainsafia**
Part of https://github.com/dotnet/aspnetcore/issues/31212.

Size of `~/github.com/dotnet/sdk/artifacts/bin/redist/Debug/dotnet/sdk/6.0.100-dev ` on this branch: 176MB
Size of `~/github.com/dotnet/sdk/artifacts/bin/redist/Debug/dotnet/sdk/6.0.100-dev ` on this main: 184MB

### CODE REVIEW

File: src/BuiltInTools/dotnet-watch/Program.cs [hunk 1]
```diff
@@ -378,5 +382,20 @@ public void Dispose()
...
         }
+
+        private static void RegisterAssemblyResolutionEvents()
+        {
+            var roslynPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "..", "..", "..", "..", "..", "..", "Roslyn", "bincore");
+
+            AssemblyLoadContext.Default.Resolving += (context, assembly) =>
+            {
+                var assemblyFilename = $"{assembly.Name}.dll";
+                if (assemblyFilename == "Microsoft.CodeAnalysis.CSharp.dll" || assemblyFilename == "Microsoft.CodeAnalysis.dll")
```
Comment (pranavkm) [inline src/BuiltInTools/dotnet-watch/Program.cs:393]: You could do the concat after you've checked the assembly name. Something like:

```C#
if (assembly.Name is "Microsoft.CodeAnalysis" or "Microsoft.CodeAnalysis.CSharp")
{
   return context.LoadFromAssemblyPath(Path.Combine(roslynPath, assemblyFilename + ".dll");
}
```

Avoids checking strings that you know would match.

File: src/BuiltInTools/dotnet-watch/Program.cs [hunk 2]
```diff
@@ -378,5 +382,20 @@ public void Dispose()
             _console.CancelKeyPress -= OnCancelKeyPress;
             _cts.Dispose();
         }
+
+        private static void RegisterAssemblyResolutionEvents()
+        {
+            var roslynPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "..", "..", "..", "..", "..", "..", "Roslyn", "bincore");
```
Comment (pranavkm) [inline src/BuiltInTools/dotnet-watch/Program.cs:388]: Could we re-use the SDK root that we calculated earlier?

File: src/RazorSdk/Tool/Program.cs [hunk 3]
```diff
@@ -49,5 +60,20 @@ public static int Main(string[] args)
 
             return result;
         }
+
+        private static void RegisterAssemblyResolutionEvents()
+        {
+            var roslynPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "..", "..", "..", "Roslyn", "bincore");
```
Comment (pranavkm) [inline src/RazorSdk/Tool/Program.cs:66]: `AppContext.BaseDirectory` instead of `AppDomain.CurrentDomain.BaseDirectory`

**Review by pranavkm — COMMENTED** [review]
Cool beans. Could you make sure hot reload works after this change (super manual testing)

File: src/RazorSdk/Tool/Program.cs (hunk 3 above)
Comment (captainsafia) [reply src/RazorSdk/Tool/Program.cs:66]: Is there a particular reason to prefer one over the other?

Comment (pranavkm) [reply src/RazorSdk/Tool/Program.cs:66]: AppDomain is considered the legacy feature and is mostly shimmed in .NET Core. AppContext is the preferred replacement.

**Review by pranavkm — APPROVED** [review]

**Pull request merged**